### PR TITLE
Fix `argmax` and `argmin` for F-order inputs

### LINKED ISSUE #4061: argmax and argmin return wrong results for F-order arrays
On CuPy master:
```python
import cupy as cp
import numpy as np


cp.core.set_routine_accelerators([])
cp.core.set_reduction_accelerators([])

a = cp.asarray([[3, 1], [4, 2]], order='F')
b = cp.asnumpy(a)
print("cp.argmax:", a.argmax(), "  np.argmax:", b.argmax())
print("cp.argmin:", a.argmin(), "  np.argmin:", b.argmin())
```
output:
```
cp.argmax: 1   np.argmax: 2
cp.argmin: 2   np.argmin: 1
```
This bug affects both CuPy's original implementation as well as the CUB block reduction (in `cupy.core._cub_reduction`), but not the CUB device reduction (`cupy.cuda.cub`). My guess is for Fortran arrays we calculated the array index (`_J`) wrong...

### PR DESCRIPTION
**Pull request by asi1024**
Depends on #4091.
Fix #4061.
Fix not to sort reduce axis of the input when `_J` is used.

### CODE REVIEW

File: cupy/core/_routines_statistics.pyx [hunk 1]
```diff
@@ -89,6 +89,9 @@ cdef ndarray _ndarray_argmax(ndarray self, axis, out, dtype, keepdims):
     for accelerator in _accelerator._routine_accelerators:
         if accelerator == _accelerator.ACCELERATOR_CUB:
             # result will be None if the reduction is not compatible with CUB
+            if self._f_contiguous and self.dtype == numpy.bool_:
+                # temporary workaround
+                self = self.astype(numpy.int8)
```
Comment (asi1024) [inline cupy/core/_routines_statistics.pyx:94]: CUB `argmax` seems to return different values to NumPy for F-order bool array inputs. In this PR, I made a temporary fix casting the inputs to int8 data type. (cc: @leofang)

**Comment by leofang** [comment]
Thanks, @asi1024! I am fixing the CUB block reduction for `argmax` and `argmin`. Do you know if NumPy always computes the indices in the C layout and then picks max/min, regard less of the actual array order being C or F?

**Comment by asi1024** [comment]
Rebased.

File: cupy/core/_routines_statistics.pyx [hunk 2]
```diff
@@ -89,6 +89,9 @@ cdef ndarray _ndarray_argmax(ndarray self, axis, out, dtype, keepdims):
     for accelerator in _accelerator._routine_accelerators:
         if accelerator == _accelerator.ACCELERATOR_CUB:
             # result will be None if the reduction is not compatible with CUB
+            if self._f_contiguous and self.dtype == numpy.bool_:
+                # temporary workaround
```
Comment (emcastillo) [inline cupy/core/_routines_statistics.pyx:93]:
```suggestion
                # temporary workaround casting the inputs to int8
                # CUB argmax seems to return different values to
                # NumPy for F-order bool array inputs
```

**Review by emcastillo — APPROVED** [review]
LGTM just the nit in the comment.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Review by emcastillo — APPROVED** [review]

**Pull request merged**